### PR TITLE
Chow ni 2023rev

### PR DESCRIPTION
**Pull request by sungeunbae**
Implementation of Chow NI tomography:
http://ds.iris.edu/ds/products/emc-nz_atom_north_chow_etal_2021_vpvs/

Details of implementation:
https://wiki.canterbury.ac.nz/display/QuakeCore/Implementation%3A+Integrating+Chow+NI+tomography+into+NZVM

The python scripts step0...step4xx.py are just one-off scripts added to the repo to be future-proof. (eg. If similar data set for South Island or a revision of this data is released). They were written last year in a separate branch, but the branch was closed without merge as we hit the dead end due to the discovery of artefact in the data set.  Luckilly when Chow's data was revised, I was able to quickly assess the data thanks to these scripts. Hence I want to include these scripts in this repo.

### CODE REVIEW

File: scripts/chow_ni_tomography/step2_map_chow_onto_ep2020_grid.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#import utm
```
Comment (joelridden) [inline scripts/chow_ni_tomography/step2_map_chow_onto_ep2020_grid.py:1]: Commented import could be removed? I don't see it being used by any commented code either

**Review by joelridden — COMMENTED** [review]
Just a few small things for tidiness

File: scripts/chow_ni_tomography/shared.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+compare_dir = work_dir / "compare"
+log_dir = work_dir / "log"
+
+MAX_DIST = 20  # km
+depth_categories = ["shallow","crust","mantle"]
+v_file_template = "surf_tomography_{}_elev{}.in"  # .format(lower(v_type), int(elev))
+v_types = ["vp", "vs", "rho"]
+
+def smart_float(n:float, afterpoint:int=2):
+    s="{:f}".format(n)
```
[the remark this answers is not in the document]
Comment (sungeunbae) [reply scripts/chow_ni_tomography/shared.py:29]: Good point. Also replaced the name of this function to p_float

**Review by joelridden — DISMISSED** [review]

File: scripts/chow_ni_tomography/step2_map_chow_onto_ep2020_grid.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+from shared import depth_categories, ep2020_yaml, NI_tomo_dir, step2_outdir, write_file, output_exists, p_float
+
+step2_outdir.mkdir(parents=True, exist_ok=True)
+
+V_types = ["vp", "vs", "rho"]
+
+"""
+This code fits Chow NI tomography onto the (lat lon) grid of EP2020 via interpolation
+"""
+def myfunc(tomo_df, mgrid_lon, mgrid_lat, elev):
```
Comment (claudio525) [inline scripts/chow_ni_tomography/step2_map_chow_onto_ep2020_grid.py:19]: better function name?

File: scripts/chow_ni_tomography/step3_combine.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            final_v_array[lat_id][lon_id] = (
+                smoothDistRatio * chow_tomo_grid_mapped_array[lat_id][lon_id]
+                + inverseRatio * ep2020_tomo_elevs_ext_array[lat_id][lon_id]
+            )
+            res = f"{lat_id},{lon_id},{smoothDistRatio:.3f},{chow_tomo_grid_mapped_array[lat_id][lon_id]:.6f},{inverseRatio:.3f},{ep2020_tomo_elevs_ext_array[lat_id][lon_id]:.6f},{final_v_array[lat_id][lon_id]:.6f}"
+            f.write(res + "\n")
+        # print(res)
+    return final_v_array
+
+def myfunc(nonzero_vert_ids,nonzero_hori_ids,boundaries,elev):
```
Comment (claudio525) [inline scripts/chow_ni_tomography/step3_combine.py:135]: better function name?

**Review by claudio525 — COMMENTED** [review]
The scripts seem a bit hacky/messy to me based on my brief look. I think it would be quite hard for anyone other than you to use these? 
Also running black over them is probably a good idea.

**Comment by sungeunbae** [comment]
@claudio525  Fair point.  We *hope* there is revision of NI tomography or release of SI tomography in the future, and this code is useful again, but we don't know if or when. 

If we ever need these scripts again, they are not meant be used off-the-shelf. Chow chaged the file format and and coordinate systems even between his revisions. So some coding effort is expected.

Main purpose of inclusion of these scripts is to show a number of steps how to merge a new tomography data in a different 3d coordinate system into the existing one, which can be quite complicated. I also documented each step in a separate wiki page to help myself/others in the future.

Without these scripts and accompanying wiki page, I think it will be *harder* for anyone to process a new tomography data. :)

**Review by claudio525 — APPROVED** [review]

**Review by joelridden — APPROVED** [review]

**Pull request merged**